### PR TITLE
add version errors

### PR DESCRIPTION
**Pull request by simonbyrne**
Extracted from #597

cc @giordano @eschnett

### CODE REVIEW

File: src/error.jl [hunk 1]
```diff
@@ -2,7 +2,30 @@ struct MPIError <: Exception
...
+end
+
+function Base.show(io::IO, err::FeatureLevelError)
+    print(io, "FeatureLevelError($(err.function_name)): Minimum MPI version is $(err.min_version)")
+end
+
+macro mpichk(expr, min_version=nothing)
+    if !isnothing(min_version) && expr.args[2].head == :tuple
+        fn = expr.args[2].args[1].value
+        if dlsym_e(Libdl.dlopen(libmpi, Libdl.RTLD_LAZY | Libdl.RTLD_GLOBAL), fn) == C_NULL
```
Comment (giordano) [inline src/error.jl:22]: Note that [`dlsym_e(...)`](https://docs.julialang.org/en/v1/stdlib/Libdl/#Base.Libc.Libdl.dlsym_e) is deprecated in favour of [`dlsym(...; throw_error=false)`](https://docs.julialang.org/en/v1/stdlib/Libdl/#Base.Libc.Libdl.dlsym)

File: src/error.jl [hunk 2]
```diff
@@ -2,7 +2,30 @@ struct MPIError <: Exception
...
+    min_version::VersionNumber # minimal MPI version required for this feature to be available
+end
+
+function Base.show(io::IO, err::FeatureLevelError)
+    print(io, "FeatureLevelError($(err.function_name)): Minimum MPI version is $(err.min_version)")
+end
+
+macro mpichk(expr, min_version=nothing)
+    if !isnothing(min_version) && expr.args[2].head == :tuple
+        fn = expr.args[2].args[1].value
```
Comment (giordano) [inline src/error.jl:21]: To be clear, I think this body is executed only the first time you call the macro, right?

Comment (simonbyrne) [reply src/error.jl:21]: It's executed when the macro expands

File: src/error.jl [hunk 3]
```diff
@@ -1,14 +1,43 @@
...
+    min_version::VersionNumber # minimal MPI version required for this feature to be available
+end
+function Base.show(io::IO, err::FeatureLevelError)
+    print(io, "FeatureLevelError($(err.function_name)): Minimum MPI version is $(err.min_version)")
+end
+
+macro mpichk(expr, min_version=nothing)
+    if !isnothing(min_version) && expr.args[2].head == :tuple
+        fn = expr.args[2].args[1].value
+        if dlsym(Libdl.dlopen(libmpi, Libdl.RTLD_LAZY | Libdl.RTLD_GLOBAL), fn; throw_error=false) == C_NULL
```
Comment (giordano) [inline src/error.jl:29]:
```suggestion
        if isnothing(dlsym(Libdl.dlopen(libmpi, Libdl.RTLD_LAZY | Libdl.RTLD_GLOBAL), fn; throw_error=false))
```

Comment (giordano) [reply src/error.jl:29]: Maybe we should keep the libmpi handle around, instead of running `dlopen` every time?  We already have it for JLL packages.

**Comment by giordano** [comment]
After this PR I'm getting
```
julia> using MPI
[ Info: Precompiling MPI [da04e1cc-30fd-572f-bb4f-1f8673147195]
WARNING: could not import System.libmpi_handle into MPI
```

**Comment by simonbyrne** [comment]
You need to use the master MPITrampoline

**Comment by simonbyrne** [comment]
I mean MPIPreferences

**Comment by giordano** [comment]
I guess this needs to be reflected in the compat section then :slightly_smiling_face:

**Pull request merged**